### PR TITLE
fix: handle empty featurized text features

### LINKED ISSUE #1667: [BUG] TrainRegressor does not throw user friendly error message if some column has only one distinct value
### SynapseML version

0.10.1

### System information

- python 3.8, scala 2.12
- Spark 3.2.1

### Describe the problem

If our training dataset has a column with only one distinct value, we'll get an error message, which does not make sense to end user, instead of it, spark ml will give precise error message about what's wrong.
_java.lang.IllegalArgumentException: **requirement failed: VectorSlicer requires that at least one feature be selected**._

### Code to reproduce issue

This is the sample dataset to reproduce this issue.
```scala
val trainDF = spark.createDataFrame(Seq(
      (0, "2", 0.50, 0.60, 0),
      (1, "2", 0.40, 0.50, 1),
      (2, "2", 0.78, 0.99, 2),
      (3, "2", 0.12, 0.34, 3),
      (0, "2", 0.50, 0.60, 0),
      (1, "2", 0.40, 0.50, 1),
      (2, "2", 0.78, 0.99, 2),
      (3, "2", 0.12, 0.34, 3),
      (0, "2", 0.50, 0.60, 0),
      (1, "2", 0.40, 0.50, 1),
      (2, "2", 0.78, 0.99, 2),
      (3, "2", 0.12, 0.34, 3)))
      .toDF("Label", "col1", "col2", "col3", "col4")
```

If we do regression model training with spark ml and prepare featurization:
```scala
import org.apache.spark.ml.feature.StringIndexer
import org.apache.spark.ml.feature.OneHotEncoder
      
val indexer = new StringIndexer().setInputCol("col1").setOutputCol("col1_index")
val indexed = indexer.fit(trainDF).transform(trainDF)

val encoder = new OneHotEncoder().setInputCols(Array("col1_index")).setOutputCols(Array("col1_ONE"))
val encoded = encoder.fit(indexed).transform(indexed)
encoded.show()
```
We'll get clear error message:
_java.lang.IllegalArgumentException:  **requirement failed: The input column col1_index should have at least two distinct values**.
  at scala.Predef$.require(Predef.scala:281)
  at org.apache.spark.ml.feature.OneHotEncoderCommon$.$anonfun$transformOutputColumnSchema$1(OneHotEncoder.scala:491)
  at scala.Option.map(Option.scala:230)
  at org.apache.spark.ml.feature.OneHotEncoderCommon$.transformOutputColumnSchema(OneHotEncoder.scala:488)
  at org.apache.spark.ml.feature.OneHotEncoderBase.$anonfun$validateAndTransformSchema$4(OneHotEncoder.scala:98)_

But if we use SynapseML TrainRegressor:
```scala
import com.microsoft.azure.synapse.ml.train._
import org.apache.spark.ml.regression.RandomForestRegressor

val regressor = new TrainRegressor()
    .setModel(new RandomForestRegressor())
    .setLabelCol("Label")

regressor.fit(trainDF)
```

the error message is misleading:
_java.lang.IllegalArgumentException: **requirement failed: VectorSlicer requires that at least one feature be selected**.
  at scala.Predef$.require(Predef.scala:281)
  at org.apache.spark.ml.feature.VectorSlicer.transformSchema(VectorSlicer.scala:146)
  at com.microsoft.azure.synapse.ml.featurize.CountSelectorModel.transformSchema(CountSelector.scala:85)_

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [ ] `area/cognitive`: Cognitive project
- [ ] `area/core`: Core project
- [ ] `area/deep-learning`: DeepLearning project
- [ ] `area/lightgbm`: Lightgbm project
- [ ] `area/opencv`: Opencv project
- [ ] `area/vw`: VW project
- [ ] `area/website`: Website
- [ ] `area/build`: Project build system
- [ ] `area/notebooks`: Samples under notebooks folder
- [ ] `area/docker`: Docker usage
- [X] `area/models`: models related issue

### What language(s) does this bug affect?

- [X] `language/scala`: Scala source code
- [X] `language/python`: Pyspark APIs
- [X] `language/r`: R APIs
- [X] `language/csharp`: .NET APIs
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/synapse`: Azure Synapse integrations
- [ ] `integrations/azureml`: Azure ML integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by ranadeepsingh**
Fixes #1667

## Problem and behavior
`CountSelectorModel` previously delegated an empty selection to Spark `VectorSlicer`, which throws `VectorSlicer requires that at least one feature be selected`. Constant or empty text can legitimately produce that state after TF-IDF/count selection.

This change:
- represents an empty selection as a nullable zero-width sparse vector with size-0 metadata;
- preserves null rows and keeps the existing `VectorSlicer` path unchanged for non-empty selections;
- lets `Featurize`, `TrainRegressor`, and the shared classifier path ignore collapsed text inputs when useful features remain;
- fails early with a column-specific message when every requested input collapses;
- preserves constant numeric and categorical behavior, feature ordering, public Params, and serialized shapes.

The exact RandomForest example from #1667 fails on unchanged `master` and passes with this patch, including nested trained-model save/load.

## Integration after #2638 and #2633
Rebased surgically onto `master` `dde02aec92eb5dfc4b9634c96bfa11ef6102315f` after #2638 and #2633 merged. The `VerifyFeaturize.scala` conflict was resolved by combining both sides rather than choosing either wholesale.

Preserved from #2638:
- `vectorAssemblerHandleInvalid` with default `skip` and `keep`/`error` modes;
- final `VectorAssembler` wiring;
- `defaultCopy` behavior;
- persistence, metadata, and public Python tests.

Preserved from #2629:
- zero-width selector handling and nullable size-0 metadata;
- direct null handling with one reusable sparse vector;
- nested empty-selector discovery and all-collapsed validation;
- RandomForest/`TrainRegressor`, schema, copy, persistence, and generated-wrapper coverage.

Added a combined regression proving collapsed text coexists with `handleInvalid=keep`, missing/NaN numeric values, `defaultCopy`, and correct one-feature metadata. Test-preservation comparison found no tests missing from either pre-rebase side. The only post-rebase compatibility adjustment sorts values in a constant-numeric assertion so it does not depend on Scala 2.12 versus 2.13 mutable-map iteration order.

## Performance and Spark compatibility
- Reuses one immutable zero-width sparse vector instead of allocating a result vector per row.
- Uses a direct null check rather than a per-row `Option` allocation.
- Adds no production RDD API, driver collection, shuffle, repartition, or unbounded materialization.
- Keeps the implementation Scala-first; generated Python wrappers expose the same behavior.

## Local validation
Validated at head `e303ba32d25c79f2297889439b4ed95700b1b15b`.

### Spark 3.5 / JDK 11
- global `compile` and `Test/compile`: passed;
- global `scalastyle` and `Test/scalastyle`: 0 findings;
- affected Scala suites: **47 passed, 0 failed, 1 intentionally ignored**:
  - `VerifyCountSelector` + `VerifyCountSelectorModel`: 11 passed;
  - `VerifyFeaturize`: 24 passed, 1 ignored;
  - `VerifyTrainRegressor`: 12 passed;
- persistence-heavy suites were rerun on Linux tmpfs to avoid unrelated WSL DrvFS commit-protocol I/O errors.

### Codegen and Python
- `core/codegen` and `core/publishLocal`: passed;
- generated `CountSelectorModel.py` and `Featurize.py`: `py_compile` passed;
- public generated-wrapper regressions from #2629 and #2638: **4 passed**;
- Black 22.3.0: **192 files unchanged**.

### Spark 4.1 compatibility replay
The current head declares merged #2638 as the compatibility prerequisite, then replays this patch onto live `upstream/spark4.1` `3364aba80e` (Spark 4.1.1, Scala 2.13.17, Java 17, Python 3.13 baseline):
- full `test:compile` compatibility gate: passed;
- `VerifyCountSelector` + `VerifyCountSelectorModel`: 11 passed;
- `VerifyFeaturize`: 24 passed, 1 ignored;
- `VerifyTrainRegressor -- -z "issue 1667"`: 3 passed;
- total: **38 passed, 0 failed, 1 intentionally ignored**.

## Review dispositions
- Suppressed performance finding: replaced per-row `Option(vector)` with a direct null check and retained null propagation.
- Suppressed test-cleanliness finding: removed the unused standalone `spark` expression.
- Exact-head review finding: assert equal score-sequence lengths before zip, preventing row-count regressions from being truncated silently.

## Current-head CI and review
- exact head: `e303ba32d25c79f2297889439b4ed95700b1b15b`;
- exact base: `dde02aec92eb5dfc4b9634c96bfa11ef6102315f`;
- rebased patch integrity: the original #2629 source patch is unchanged; the only new commit declares #2638 for release replay;
- Azure build `231604593` exposed stale, already-synced prerequisites; the current head removes those entries, retains only required #2638, and passes the exact Spark 4.1 replay plus full `test:compile`;
- GitHub checks and exact-head Copilot review are rerunning after the push;
- the previous Azure run belongs to the superseded head and is not readiness evidence; fresh pull-request validation is being queued for this head.
## Overlap / merge risk
#2638 and #2633 are now part of the base. Their behavior and tests are incorporated and validated above, so there is no remaining merge-order dependency between those merged PRs and #2629. A stale branch that still carries pre-merge edits to `Featurize.scala` or `VerifyFeaturize.scala` may still need a normal conflict resolution against current `master`.

### CODE REVIEW

File: core/src/main/scala/com/microsoft/azure/synapse/ml/featurize/CountSelector.scala [hunk 1]
```diff
@@ -75,15 +77,37 @@ class CountSelectorModel(val uid: String) extends Model[CountSelectorModel]
...
 
+  private def hasEmptySelection: Boolean = getIndices.isEmpty
+
+  private def emptyOutputMetadata: Metadata = new AttributeGroup(getOutputCol, 0).toMetadata()
+
+  private val emptyVector = udf { vector: Vector =>
+    Option(vector)
+      .map(_ => Vectors.sparse(0, Array.empty[Int], Array.empty[Double]))
+      .orNull
+  }
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply core/src/main/scala/com/microsoft/azure/synapse/ml/featurize/CountSelector.scala]: Fixed in 22ef360382: the empty sparse vector is now created once per transform/UDF closure and reused for every non-null row, preserving null propagation.

File: core/src/test/scala/com/microsoft/azure/synapse/ml/featurize/VerifyCountSelector.scala [hunk 2]
```diff
@@ -40,6 +60,36 @@ class VerifyCountSelector extends EstimatorFuzzing[CountSelector] with VerifyCou
...
+      .setOutputCol("selected")
+
+    val copiedModel = model.copy(ParamMap.empty)
+    assert(copiedModel.getIndices.isEmpty)
+
+    val modelDir = Paths.get("core", "target", "issue-1667", "count-selector-empty-model").toFile
+    if (modelDir.exists()) {
+      FileUtils.deleteDirectory(modelDir)
+    }
+
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply core/src/test/scala/com/microsoft/azure/synapse/ml/featurize/VerifyCountSelector.scala]: Fixed in 22ef360382: the persistence test now uses TestBase.tmpDir for isolated per-run storage and retains cleanup.

File: core/src/test/scala/com/microsoft/azure/synapse/ml/train/VerifyTrainRegressor.scala [hunk 3]
```diff
@@ -51,6 +51,89 @@ class VerifyTrainRegressor extends EstimatorFuzzing[TrainRegressor] {
...
+      .toSeq
+    val scoresWithConstantText = modelWithConstantText.transform(dataset)
+      .select("prediction")
+      .collect()
+      .map(_.getDouble(0))
+      .toSeq
+
+    baselineScores.zip(scoresWithConstantText).foreach { case (expected, actual) =>
+      assert(math.abs(expected - actual) < 1e-8)
+    }
```
[the remark this answers is not in the document]
Comment (ranadeepsingh) [reply core/src/test/scala/com/microsoft/azure/synapse/ml/train/VerifyTrainRegressor.scala:88]: Fixed in 9351894030: assert the baseline and constant-text score sequences have equal lengths before comparing paired predictions, so future row dropping cannot be hidden by zip truncation. JDK 11 VerifyTrainRegressor: 12/12 passed; final Spark 4.1 issue-1667 replay: 3/3 passed.

**Comment by ranadeepsingh** [comment]
Current-head gate audit: Copilot reviewed 9351894030 (6/6 files, 0 new comments); all 3 threads are resolved and there are no current-head suppressed findings. Azure build 231567218 is blocked before PR replay by release-compat prerequisite 04897bae9baa08f0d67855566f7bad235791d508 conflicting only in unrelated cognitive Search files on spark4.1. The exact #2638 + #2629 replay on upstream/spark4.1@3364aba80e independently passed core/Test/compile and 38 targeted tests (1 intentional ignore). Final Get-PrReadiness was run; the PR remains blocked until the shared spark4.1 prerequisite/sync gate is repaired and Azure is rerun.
